### PR TITLE
Allow limited access to underlying number and safety status of EncryptedNumber

### PR DESCRIPTION
**Pull request by mhsiah**
This pull request allows the calling context to access the ciphertext of this EncryptedNumber as a byte array. The purpose is to allow passing said array to a native library for processing, thus allowing the use of implementations of arithmetic other than the one provided with the JDK. The changes described are a stop-gap and are expected to be obviated by a future redesign of the EncryptedNumber interface.

### CODE REVIEW

File: src/main/java/com/n1analytics/paillier/EncryptedNumber.java [hunk 1]
```diff
@@ -416,4 +416,28 @@ public boolean equals(EncryptedNumber o) {
             context.equals(o.context) &&
             ciphertext.equals(o.ciphertext));
   }
+
+  /**
+   * Return a byte array representing the underlying raw number.
```
Comment (wilko77) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: 'underlying raw number' is ambiguous (for me, at least). Better: 'ciphertext'.

File: src/main/java/com/n1analytics/paillier/EncryptedNumber.java [hunk 2]
```diff
@@ -416,4 +416,28 @@ public boolean equals(EncryptedNumber o) {
...
+   *
+   * WARNING: This method may return unencrypted information! Under no
+   * circumstances should the return value be sent anywhere.
```
Comment (wilko77) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: Also add comment on how one should access the ciphertext instead.

File: src/main/java/com/n1analytics/paillier/EncryptedNumber.java [hunk 3]
```diff
@@ -416,4 +416,28 @@ public boolean equals(EncryptedNumber o) {
...
+  }
+
+  /**
+   * Return true or false according to whether this EncryptedNumber is
+   * safe for transmission.
+   *
+   * WARNING: This method is not intended for general use and may be
+   * removed in a future release. Think twice before using it.
+   */
+  public boolean getIsSafe() {
```
Comment (wilko77) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: I would name the getter for a boolean value with an 'is' prefix. Instead of getIsSafe() -> isSafe().

Comment (mhsiah) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: So would I, but the variable is already called _isSafe_, so there'd be a name conflict.

File: src/main/java/com/n1analytics/paillier/EncryptedNumber.java (hunk 1 above)
Comment (mhsiah) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: Fair enough. FWIW the vagueness was a deliberate part of the trade-off between documenting the function but making it "scary" and discouraging its use.

File: src/main/java/com/n1analytics/paillier/EncryptedNumber.java (hunk 3 above)
Comment (maxott) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: Are you needing the `getIsSafe` for anything else but serialising it for use in a JNI library? If yes, it's an entirely internal process and I'm wondering if a very targeted method may be more beneficial. There may be concerns about 'polluting' this library with very specific 'external' requirements, but this is a performance concern and may be acceptable for those reasons

Comment (mhsiah) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: At the moment I 

- take an array of EncryptedNumbers,
- unpack the ciphertexts,
- put them in a big byte array,
- send that to the GPU,
- have the GPU do its thing,
- retrieve the result from the GPU as a byte array,
- build the resulting ciphertexts, then
- _recreate the array of EncryptedNumbers_ using the new ciphertexts,
- return this new array as the result.

Regarding the last step, the EncryptedNumber constructor takes an _isSafe_ parameter which is _false_ by default (as it should be), so if I just recreate the EncryptedNumbers with the default, all the ones that were already safe will be needlessly (and expensively) re-encrypted. Since the calculation I just did on the GPU does not change the "safety" (in the case of the `inner` and `multiply` methods of DenseEncryptedVector), I obtain the original isSafe parameter using this `getIsSafe` method, and pass it to the constructor of EncryptedNumber when I construct the corresponding result.

Comment (maxott) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: Thanks for the clarification. What I meant by 'internal' process is that you are doing all this to improve the performance of the library itself. You are not wanting these additional APIs to enhance the usage of the library. In other words, what you need should ideally not even be made aware to a 'normal' user. 

Therefore, my concerns go even further. You mention that you primarily deal with arrays of encrypted numbers, so I assume that you may serialise the entire vector into a single byte array which you then hand down to your GPU library. If you do this, you may actually create a a few intermittent byte arrays and a few additional copies between all them. In addition, I wonder how much extra 'massaging' you need to do in your GPU library to properly interpret the big integer serialisation the Java library performs in it's `toByteArray`. 

In other words, this entire process seems to be rather inefficient and kind of counter-productive to the overall goal you have - make this bird fly faster. And given all the effort you already put into the GPU library, I wonder if we shouldn't spend a bit more effort on the passing back and forth across the JNI boundary.

Now it's perfectly possible, that I completely misjudging that overhead

Comment (mhsiah) [inline src/main/java/com/n1analytics/paillier/EncryptedNumber.java]: Right, so the serialisation you refer to occurs at step 3 (and in reverse at step 7) in the list I posted above. Indeed I am obliged to create at least (but I think _only_) two intermediate arrays to "serialise" the `BigInteger[]` for consumption by the GPU. Happily there isn't much in the way of 'massaging' to do, since the only oddities of the `BigInteger.toByteArray()` result are (i) it is big endian, and (ii) the "most-significant" byte in the resulting array is allowed to be zero (this is a weird normalisation, for handling signed BigIntegers I think).

I have measured the overhead of this array creation/massaging/copying at various times in the past and it has never been significant compared to the the other calculations, though this may have changed recently with the incorporation of Felix's trick, which has invalidated most of my previous performance numbers. With some extra work I believe I could reduce the serialisation I do to essentially a single copy from the JVM to the GPU memory, if this turns out to be a bottleneck.

The performance question is still more-or-less orthogonal to the problem that this PR was supposed to solve though, namely: at some point, somehow, I need to get access to the `ciphertext`s inside the `EncryptedNumber[]` so that I can send them to the GPU for processing.

**Comment by mhsiah** [comment]
Following the discussion above among others we have agreed that this needs more thought.
